File: ts/src/snapshot.ts
import { createHash } from "node:crypto";
import { type Dirent, readFileSync, readdirSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";
import { ContainmentError } from "./errors.js";
import { NodeId } from "./ids.js";
import { RESERVED_NAMESPACE, isPortableRelativePath, readJson, writeJsonAtomic } from "./paths.js";
import { SearchIndex, compareCodepoints } from "./search.js";
import { VectorIndex } from "./similarity.js";
import { Index } from "./structural-index.js";

export const SNAPSHOT_SCHEMA_VERSION = 1;
export const SNAPSHOT_LANG = "ts";
export const SNAPSHOT_REL_PATH = `${RESERVED_NAMESPACE}/snapshot.ts.json`;
export { readJson, writeJsonAtomic };

const SHA256_RE = /^[0-9a-f]{64}$/;
const SNAPSHOT_KEYS = ["version", "lang", "manifest", "structural", "search", "vectors"];
const MANIFEST_ROW_KEYS = ["path", "sha256", "uid"];

export function snapshotPath(root: string): string {
  return join(root, SNAPSHOT_REL_PATH);
}

export function hashBytes(data: Buffer | Uint8Array): string {
  return createHash("sha256").update(data).digest("hex");
}

export interface CorpusFile {
  readonly path: string; // root-relative POSIX
  readonly data: Buffer;
  readonly sha256: string;
}

export interface CorpusFileStat {
  readonly path: string;
  readonly mtimeMs: number;
  readonly size: number;
}

export interface CorpusFingerprint {
  readonly files: readonly CorpusFileStat[];
}

interface WalkedCorpusPath {
  readonly path: string;
  readonly fullPath: string;
}

/** Root-relative POSIX path (forward slashes on every platform), the cross-language form. */
function relPosix(root: string, full: string): string {
  return relative(root, full).split(sep).join("/");
}

function listCorpusMarkdownPaths(root: string): WalkedCorpusPath[] {
  const files: WalkedCorpusPath[] = [];
  const walk = (dir: string, atRoot: boolean): void => {
    const entries: Dirent[] = readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      const full = join(dir, entry.name);
      if (entry.isSymbolicLink()) continue;
      if (entry.isDirectory()) {
        if (atRoot && entry.name === RESERVED_NAMESPACE) continue;
        walk(full, false);
      } else if (entry.isFile() && entry.name.endsWith(".md")) {
        files.push({ path: relPosix(root, full), fullPath: full });
      }
    }
  };
  walk(root, true);
  files.sort((a, b) => compareCodepoints(a.path, b.path));
  return files;
}

/** Byte-level walk: read each .md file's bytes once and hash them. Skips the private
 * `.nodes-index` tree, symlinks, and non-files. Sorted by root-relative POSIX path so the
 * order matches Python's `sorted(root.rglob("*.md"))`. */
export function iterCorpusFiles(root: string): CorpusFile[] {
  return listCorpusMarkdownPaths(root).map(({ path, fullPath }) => {
    const data = readFileSync(fullPath);
    return { path, data, sha256: hashBytes(data) };
  });
}

/** Stat-level walk over the same corpus file set as `iterCorpusFiles`. Does not read file bodies. */
export function listCorpusFileStats(root: string): CorpusFileStat[] {
  return listCorpusMarkdownPaths(root).map(({ path, fullPath }) => {
    const stat = statSync(fullPath);
    return { path, mtimeMs: stat.mtimeMs, size: stat.size };
  });
}

/** Cheap external-change fingerprint for resident consumers. Not a content-identity hash. */
export function readCorpusFingerprint(root: string): CorpusFingerprint {
  return { files: listCorpusFileStats(root) };
}

export function sameCorpusFingerprint(a: CorpusFingerprint, b: CorpusFingerprint): boolean {
  if (a.files.length !== b.files.length) return false;
  for (let i = 0; i < a.files.length; i++) {
    const left = a.files[i];
    const right = b.files[i];
    if (left.path !== right.path || left.mtimeMs !== right.mtimeMs || left.size !== right.size) return false;
  }
  return true;
}

export interface ManifestEntry {
  readonly path: string;
  readonly sha256: string;
  readonly uid: string;
}

export interface Snapshot {
  manifest: ManifestEntry[];
  index: Index;
  searchIndex: SearchIndex;
  vectorIndex: VectorIndex | null;
}

export function pathForNodeId(nodeId: string): string {
  const nid = NodeId.parse(nodeId);
  return `${nid.kind}/${nid.slug.replace(/:/g, "__")}.md`;
}

export function writeSnapshot(
  root: string,
  manifest: ManifestEntry[],
  index: Index,
  searchIndex: SearchIndex,
  vectorIndex: VectorIndex | undefined,
): void {
  const doc = {
    version: SNAPSHOT_SCHEMA_VERSION,
    lang: SNAPSHOT_LANG,
    manifest: manifest.map((m) => ({ path: m.path, sha256: m.sha256, uid: m.uid })),
    structural: index.toDict(),
    search: searchIndex.toDict(),
    vectors: vectorIndex !== undefined ? vectorIndex.toDict() : null,
  };
  writeJsonAtomic(root, SNAPSHOT_REL_PATH, doc);
}

function validateManifestPath(path: string): void {
  if (!isPortableRelativePath(path) || path.split("/", 1)[0] === RESERVED_NAMESPACE) {
    throw new Error("snapshot manifest row path must be a portable root-relative .md path");
  }
}

function parseManifest(raw: unknown): ManifestEntry[] {
  if (!Array.isArray(raw)) throw new Error("snapshot manifest is not an array");
  const entries: ManifestEntry[] = [];
  for (const e of raw) {
    if (typeof e !== "object" || e === null) throw new Error("snapshot manifest row is not an object");
    const row = e as Record<string, unknown>;
    for (const key of MANIFEST_ROW_KEYS) {
      if (!(key in row)) throw new Error(`snapshot manifest row missing ${key}`);
    }
    const { path, sha256, uid } = row;
    if (typeof path !== "string") throw new Error("snapshot manifest row path must be a string");
    validateManifestPath(path);
    if (typeof sha256 !== "string") throw new Error("snapshot manifest row sha256 must be a string");
    if (!SHA256_RE.test(sha256)) throw new Error("snapshot manifest row sha256 must be 64 lowercase hex chars");
    if (typeof uid !== "string") throw new Error("snapshot manifest row uid must be a string");
    entries.push({ path, sha256, uid });
  }
  if (new Set(entries.map((m) => m.uid)).size !== entries.length) throw new Error("snapshot manifest: duplicate uid");
  if (new Set(entries.map((m) => m.path)).size !== entries.length) throw new Error("snapshot manifest: duplicate path");
  return entries;
}

function setsEqual(a: Set<string>, b: Set<string>): boolean {
  return a.size === b.size && [...a].every((x) => b.has(x));
}

function mapsEqual(a: Map<string, string>, b: Map<string, string>): boolean {
  if (a.size !== b.size) return false;
  for (const [k, v] of a) if (b.get(k) !== v) return false;
  return true;
}

/** Reads and validates ONLY the cache file. Returns null for any cache problem (missing file,
 * invalid JSON, version/lang mismatch, integrity failure, embedder-configured vector mismatch).
 * Never parses corpus files, so it can never raise a corpus error — any throw here is a cache
 * problem and resolves to a silent full rebuild upstream. */
export function loadSnapshot(root: string, embedderNamespace: string | null): Snapshot | null {
  try {
    const doc = readJson(root, SNAPSHOT_REL_PATH);
    if (doc === null) return null;
    if (typeof doc !== "object") return null;
    const d = doc as Record<string, unknown>;
    for (const key of SNAPSHOT_KEYS) {
      if (!(key in d)) throw new Error(`snapshot document missing ${key}`);
    }
    if (d.version !== SNAPSHOT_SCHEMA_VERSION || d.lang !== SNAPSHOT_LANG) return null;

    const manifest = parseManifest(d.manifest);
    const manifestUids = new Set(manifest.map((m) => m.uid));

    const index = Index.fromDict(d.structural);
    if (!setsEqual(new Set(index.byUid.keys()), manifestUids)) return null;
    const expectedIds = new Map<string, string>();
    for (const [uid, entry] of index.byUid) expectedIds.set(uid, entry.id);
    for (const m of manifest) {
      if (m.path !== pathForNodeId(expectedIds.get(m.uid) as string)) {
        throw new Error("snapshot manifest path does not match structural id");
      }
    }

    const searchIndex = SearchIndex.fromDict(d.search);
    if (!setsEqual(new Set(searchIndex.lengths.keys()), manifestUids)) return null;
    if (!mapsEqual(searchIndex.idByUid, expectedIds)) return null;

    let vectorIndex: VectorIndex | null = null;
    if (embedderNamespace !== null) {
      const vec = d.vectors;
      if (typeof vec !== "object" || vec === null) return null;
      if ((vec as Record<string, unknown>).namespace !== embedderNamespace) return null;
      vectorIndex = VectorIndex.fromDict(vec);
      if (!setsEqual(new Set(vectorIndex.vectors.keys()), manifestUids)) return null;
      if (!mapsEqual(vectorIndex.idByUid, expectedIds)) return null;
    }

    return { manifest, index, searchIndex, vectorIndex };
  } catch (e) {
    if (e instanceof ContainmentError) throw e;
    // loadSnapshot only ever reads the cache file, so any thrown Error is a cache problem
    // (absent/locked file, malformed JSON, failed integrity check) -> rebuild. This is the
    // closest faithful mirror of Python's `except (OSError, ValueError)`: every cache-unusable
    // signal here surfaces as an Error subclass, while JS's error taxonomy gives no class-based
    // way to separate them further. A non-Error throw is not a cache signal — rethrow it.
    if (!(e instanceof Error)) throw e;
    return null;
  }
}

File: python/src/nodes/core/snapshot.py
from __future__ import annotations

import hashlib
import os
import re
from dataclasses import dataclass
from pathlib import Path

from nodes.core.paths import (
    RESERVED_NAMESPACE,
    is_portable_relative_path,
    path_for_node_id,
    read_json as read_json,
    write_json_atomic as write_json_atomic,
)
from nodes.core.structural_index import Index
from nodes.core.search import SearchIndex
from nodes.core.similarity import VectorIndex

SNAPSHOT_SCHEMA_VERSION = 3
SNAPSHOT_LANG = "py"
SNAPSHOT_REL_PATH = f"{RESERVED_NAMESPACE}/snapshot.py.json"
_SHA256_RE = re.compile(r"[0-9a-f]{64}")
_SNAPSHOT_KEYS = frozenset({"version", "lang", "manifest", "structural", "search", "vectors"})
_MANIFEST_ROW_KEYS = frozenset({"path", "sha256", "uid"})


def snapshot_path(root: Path | str) -> Path:
    return Path(root) / SNAPSHOT_REL_PATH


def hash_bytes(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest()


@dataclass(frozen=True)
class CorpusFile:
    path: str  # root-relative POSIX
    data: bytes
    sha256: str


def iter_corpus_files(root: Path | str) -> list[CorpusFile]:
    """Walk regular Markdown files without following symlinks or hiding failures."""
    root = Path(root)
    files: list[CorpusFile] = []

    def walk(directory: Path, rel_parts: tuple[str, ...]) -> None:
        with os.scandir(directory) as entries:
            for entry in entries:
                if entry.is_symlink():
                    continue
                if entry.is_dir(follow_symlinks=False):
                    if not rel_parts and entry.name == RESERVED_NAMESPACE:
                        continue
                    walk(Path(entry.path), (*rel_parts, entry.name))
                elif entry.is_file(follow_symlinks=False) and entry.name.endswith(".md"):
                    data = Path(entry.path).read_bytes()
                    files.append(CorpusFile(path="/".join((*rel_parts, entry.name)), data=data, sha256=hash_bytes(data)))

    walk(root, ())
    files.sort(key=lambda f: f.path)
    return files


@dataclass(frozen=True)
class ManifestEntry:
    path: str
    sha256: str
    uid: str


@dataclass
class Snapshot:
    manifest: list[ManifestEntry]
    index: Index
    search_index: SearchIndex
    vector_index: VectorIndex | None


def write_snapshot(
    root: Path | str,
    manifest: list[ManifestEntry],
    index: Index,
    search_index: SearchIndex,
    vector_index: VectorIndex | None,
) -> None:
    doc = {
        "version": SNAPSHOT_SCHEMA_VERSION,
        "lang": SNAPSHOT_LANG,
        "manifest": [{"path": m.path, "sha256": m.sha256, "uid": m.uid} for m in manifest],
        "structural": index.to_dict(),
        "search": search_index.to_dict(),
        "vectors": vector_index.to_dict() if vector_index is not None else None,
    }
    write_json_atomic(root, SNAPSHOT_REL_PATH, doc)


def _parse_manifest(raw: object) -> list[ManifestEntry]:
    if not isinstance(raw, list):
        raise ValueError("snapshot manifest is not a list")
    entries = []
    for e in raw:
        if not isinstance(e, dict):
            raise ValueError("snapshot manifest row is not a dict")
        missing = _MANIFEST_ROW_KEYS - e.keys()
        if missing:
            raise ValueError(f"snapshot manifest row missing {sorted(missing)[0]}")
        path = e["path"]
        sha256 = e["sha256"]
        uid = e["uid"]
        if not isinstance(path, str):
            raise ValueError("snapshot manifest row path must be a string")
        _validate_manifest_path(path)
        if not isinstance(sha256, str):
            raise ValueError("snapshot manifest row sha256 must be a string")
        if _SHA256_RE.fullmatch(sha256) is None:
            raise ValueError("snapshot manifest row sha256 must be 64 lowercase hex chars")
        if not isinstance(uid, str):
            raise ValueError("snapshot manifest row uid must be a string")
        entries.append(ManifestEntry(path=path, sha256=sha256, uid=uid))
    uids = [e.uid for e in entries]
    paths = [e.path for e in entries]
    if len(set(uids)) != len(uids):
        raise ValueError("snapshot manifest: duplicate uid")
    if len(set(paths)) != len(paths):
        raise ValueError("snapshot manifest: duplicate path")
    return entries


def _validate_manifest_path(path: str) -> None:
    if not is_portable_relative_path(path) or path.split("/", 1)[0] == RESERVED_NAMESPACE:
        raise ValueError("snapshot manifest row path must be a portable root-relative .md path")


def load_snapshot(root: Path | str, embedder_namespace: str | None) -> Snapshot | None:
    try:
        doc = read_json(root, SNAPSHOT_REL_PATH)
        if doc is None:
            return None
        if not isinstance(doc, dict):
            return None
        missing = _SNAPSHOT_KEYS - doc.keys()
        if missing:
            raise ValueError(f"snapshot document missing {sorted(missing)[0]}")
        if doc.get("version") != SNAPSHOT_SCHEMA_VERSION or doc.get("lang") != SNAPSHOT_LANG:
            return None

        manifest = _parse_manifest(doc["manifest"])
        manifest_uids = {m.uid for m in manifest}

        index = Index.from_dict(doc["structural"])
        if set(index.by_uid) != manifest_uids:
            return None
        expected_ids = {uid: entry.id for uid, entry in index.by_uid.items()}
        for m in manifest:
            if m.path != path_for_node_id(expected_ids[m.uid]):
                raise ValueError("snapshot manifest path does not match structural id")

        search_index = SearchIndex.from_dict(doc["search"])
        if set(search_index.lengths) != manifest_uids:
            return None
        if search_index.id_by_uid != expected_ids:
            return None

        vector_index: VectorIndex | None = None
        if embedder_namespace is not None:
            vec = doc.get("vectors")
            if not isinstance(vec, dict):
                return None
            if vec.get("namespace") != embedder_namespace:
                return None
            vector_index = VectorIndex.from_dict(vec)
            if set(vector_index.vectors) != manifest_uids:
                return None
            if vector_index.id_by_uid != expected_ids:
                return None

        return Snapshot(
            manifest=manifest,
            index=index,
            search_index=search_index,
            vector_index=vector_index,
        )
    except (OSError, ValueError):
        return None
